File: ruby/zork/game_state.rb
require "./game_map"
require "./player"

$explore_regex = /^EXPLORE (?<direction>NORTH|EAST|SOUTH|WEST)$/i;
$move_regex = /^MOVE (?<direction>NORTH|EAST|SOUTH|WEST)$/i;
$sys_regex = /^EXIT$/i;

class GameState
  attr_accessor :map, :player, :last_command

  def initialize
    @map = GameMap.new
    @player = Player.new
    @player.current_room = @map.starting_room
  end

  def handle_commands
    @player.status
    exits = @player.current_room.exits.map { |sym| sym.to_s.upcase }
    expeditions = @player.current_room.expeditions.map { |sym| sym.to_s.upcase }
    puts "You may MOVE #{exits.join('/')}." unless exits.empty?
    puts "You may EXPLORE #{expeditions.join('/')}." unless expeditions.empty?

    next_command = nil
    loop do
      print "What is your command? "
      next_command = gets.chomp
      break if next_command.match?($move_regex) or next_command.match?($explore_regex) or next_command.match?($sys_regex)
      puts "That is an invalid command."
    end

    @last_command = next_command
    case last_command
    when $move_regex
      @player.current_room = @player.current_room.rooms[$~[:direction].downcase.intern] if exits.include?($~[:direction].upcase)
    when $explore_regex
      @player.current_room = @map.generate_unexplored_room(@player.current_room, $~[:direction].downcase.intern)
    when $sys_regex
      exit
    end
  end
end

File: lua/zork/game_state.lua
local utils = require "utils"
local GameMap = require "game_map"
local Player = require "player"

local directions = { "NORTH", "EAST", "SOUTH", "WEST" }

local explore_command = "EXPLORE (%a+)"
local move_command = "MOVE (%a+)"
local system_command = "EXIT"

GameState = {}
function GameState:new(game_map, player)
   local game_state = Class.new(self, {
      game_map = game_map,
      player = player,
      last_command = nil
   })

   game_state.player.current_room = game_map.starting_room
   return game_state
end

function GameState:handle_commands()
   self.player:status()
   self.player.current_room:get_exits()

   local next_input = nil
   local next_direction = nil
   repeat
      print("What is your command? ")
      _, next_input = pcall(io.read)
      _, next_input = pcall(utils.trim, next_input)
      _, next_input = pcall(string.upper, next_input)
   until next_input ~= nil

   next_direction = string.match(next_input, move_command)
   if next_direction ~= nil then
      self.player.current_room = self.player.current_room.rooms[next_direction]
   end

   next_direction = string.match(next_input, explore_command)
   if next_direction ~= nil then
      self.player.current_room = self.game_map:generate_unexplored_room(self.player.current_room, next_direction)
   end

   if string.match(next_input, system_command) ~= nil then
      os.exit(0)
   end

   self.last_command = next_input
end

return GameState
   
